server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java

```diff
@@ -52,17 +61,79 @@ public Port getPort() {
+                    (endpoint, route) -> {
+                        if (route.size() == 1) {
+                            JMAPRoute next = route.iterator().next();
+                            switch (endpoint.getVerb()) {
+                                case GET:
+                                    routes.get(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                                case POST:
+                                    routes.post(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                                case DELETE:
+                                    routes.delete(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                                case OPTIONS:
+                                    routes.options(endpoint.getPath(), (req, res) -> next.getAction().apply(req, res));
+                                    break;
+                            }
+                        } else if (route.size() == 2) {
+                            ImmutableList<JMAPRoute> sorted = route.stream()
+                                .sorted(Comparator.comparing(JMAPRoute::getVersion))
+                                .collect(Guavate.toImmutableList());
+                            JMAPRoute draftRoute = sorted.get(0);
+                            JMAPRoute rfc8621Route = sorted.get(1);
+
+                            switch (endpoint.getVerb()) {
+                                case GET:
+                                    routes.get(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                                case POST:
+                                    routes.post(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                                case DELETE:
+                                    routes.delete(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                                case OPTIONS:
+                                    routes.options(endpoint.getPath(), (req, res) -> chooseVersionRoute(req, res, draftRoute, rfc8621Route));
+                                    break;
+                            }
+                        }
+                    }))
+                )
                 .wiretap(configuration.wiretapEnabled())
                 .bindNow());
         }
     }
 
+    private Publisher<Void> chooseVersionRoute(HttpServerRequest request, HttpServerResponse response, JMAPRoute draftRoute, JMAPRoute rfc8621Route) {
+        if (hasRfc8621AcceptHeader(request)) {
+            return rfc8621Route.getAction().apply(request, response);
+        }
```

Comment:
Actually I'm getting stuck in an endless loop by trying to fix all those comments... If I want to throw back a 400 error response here, I need a response (making then next comment not doable... and I keep hitting issues like this)

Reply:
PR comments are not an absolute truth, and there might be some conflicting objective.

Stay cold headed, and make choices ^^

Reply:
IMO you are mixing choosing a route & applying it.

If you return an `Action`, you will be aable to:
 - return the action of the chose route
 - return `(req, res) -> res.fail400()` in case you wonna fail

Take a decision. Don't apply it here.

IMO it will makes things more readable. (and actually presented like this it do not conflict)